### PR TITLE
osd/PG: restart recovery if NotRecovering and unfound found

### PR DESCRIPTION
**Pull request by liewegas**
If we are in recovery_unfound state waiting for unfound objects, and we
find them, we need to restart the recovery reservation process so that we
can recover.  Do this by queueing DoRecover() event instead of calling
queue_recovery() (which won't do anything since we're not in
recoverying|backfilling pg states).

Make the parent Active state ignore DoRecovery so that if we are already
in some phase of recovery/backfill the event gets ignored.  It is already
handled by the other important substates that care, like Clean (for
repair's benefit).

I'm not sure why states like Activating are paying attention tot his vevent...

Fixes: http://tracker.ceph.com/issues/22145
Signed-off-by: Sage Weil <sage@redhat.com>

### CODE REVIEW

File: qa/suites/rados/singleton-nomsgr/all/recovery-unfound-found.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      count: 2
+      size: 20 # GB
+tasks:
+- install:
+- ceph:
+    fs: xfs
+    conf:
+      osd:
+        osd recovery sleep: .1
+	osd objectstore: filestore
```
Comment (badone) [inline qa/suites/rados/singleton-nomsgr/all/recovery-unfound-found.yaml]: Did a tab slip in here?

Comment (liewegas) [reply qa/suites/rados/singleton-nomsgr/all/recovery-unfound-found.yaml]: yep, fixed!

Comment (badone) [reply qa/suites/rados/singleton-nomsgr/all/recovery-unfound-found.yaml]: Good to go

**Comment by xiexingguo** [comment]
> I'm not sure why states like Activating are paying attention tot his vevent...

+1

File: qa/suites/rados/upgrade/jewel-x-singleton/o [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[HANDLER_OUTPUT]
```
Comment (tchaikov) [inline qa/suites/rados/upgrade/jewel-x-singleton/o]: @liewegas teuthology-suite panics at seeing this file, as it is not ended with  `.yaml` and this directory will be empty without this file. which renders the test matrix representing this directory an empty one:
```
Traceback (most recent call last):
  File "/home/kchai/teuthology/virtualenv/bin/teuthology-suite", line 11, in <module>
    load_entry_point('teuthology', 'console_scripts', 'teuthology-suite')()
  File "/home/kchai/teuthology/scripts/suite.py", line 137, in main
    return teuthology.suite.main(args)
  File "/home/kchai/teuthology/teuthology/suite/__init__.py", line 88, in main
    run.prepare_and_schedule()
  File "/home/kchai/teuthology/teuthology/suite/run.py", line 309, in prepare_and_schedule
    num_jobs = self.schedule_suite()
  File "/home/kchai/teuthology/teuthology/suite/run.py", line 476, in schedule_suite
    build_matrix(suite_path, subset=self.args.subset)
  File "/home/kchai/teuthology/teuthology/suite/build_matrix.py", line 50, in build_matrix
    mat, first, matlimit = _get_matrix(path, subset)
  File "/home/kchai/teuthology/teuthology/suite/build_matrix.py", line 60, in _get_matrix
    mat = _build_matrix(path, mincyclicity=outof)
  File "/home/kchai/teuthology/teuthology/suite/build_matrix.py", line 122, in _build_matrix
    fn)
  File "/home/kchai/teuthology/teuthology/suite/build_matrix.py", line 122, in _build_matrix
    fn)
  File "/home/kchai/teuthology/teuthology/suite/build_matrix.py", line 131, in _build_matrix
    return matrix.Sum(item, submats)
  File "/home/kchai/teuthology/teuthology/suite/matrix.py", line 225, in __init__
    "Sum requires non-empty _submats"
AssertionError: Sum requires non-empty _submats
```

Comment (liewegas) [reply qa/suites/rados/upgrade/jewel-x-singleton/o]: fixed!

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-11-17_09:57:45-rados-wip-kefu-testing-2017-11-17-1613-distro-basic-smithi/

quite a few tests failed, with backtrace like
```
ceph version 13.0.0-3301-gc1e33f9 (c1e33f94784096fb4fd6761b8ff933c732ffaf3a) mimic (dev)
 1: (()+0xafc1a4) [0x55fd474e51a4]
 2: (()+0x11390) [0x7fb3f1635390]
 3: (gsignal()+0x38) [0x7fb3f05d0428]
 4: (abort()+0x16a) [0x7fb3f05d202a]
 5: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x28e) [0x55fd475283fe]
 6: (PG::RecoveryState::Crashed::Crashed(boost::statechart::state<PG::RecoveryState::Crashed, PG::RecoveryState::RecoveryMachine, boost::mpl:
:list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_:
:na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::my_context)+0x135) [0x55fd46fba1e5]
 7: (()+0x612e66) [0x55fd46ffbe66]
 8: (boost::statechart::simple_state<PG::RecoveryState::Primary, PG::RecoveryState::Started, PG::RecoveryState::Peering, (boost::statechart::
history_mode)0>::react_impl(boost::statechart::event_base const&, void const*)+0x16e) [0x55fd470383be]
 9: (boost::statechart::simple_state<PG::RecoveryState::Active, PG::RecoveryState::Primary, PG::RecoveryState::Activating, (boost::statechart
::history_mode)0>::react_impl(boost::statechart::event_base const&, void const*)+0x1c1) [0x55fd470363d1]
 10: (boost::statechart::simple_state<PG::RecoveryState::WaitRemoteRecoveryReserved, PG::RecoveryState::Active, boost::mpl::list<mpl_::na, mp
l_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl
_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::react_impl(boost::statechart::event_base const&, void co
nst*)+0x73) [0x55fd47035323]
 11: (boost::statechart::state_machine<PG::RecoveryState::RecoveryMachine, PG::RecoveryState::Initial, std::allocator<void>, boost::statechart::null_exception_translator>::process_event(boost::statechart::event_base const&)+0x69) [0x55fd4700f579]
 12: (PG::process_peering_event(PG::RecoveryCtx*)+0x455) [0x55fd46ff6255]
 13: (OSD::process_peering_events(std::__cxx11::list<PG*, std::allocator<PG*> > const&, ThreadPool::TPHandle&)+0x48d) [0x55fd46f264bd]
 14: (ThreadPool::BatchWorkQueue<PG>::_void_process(void*, ThreadPool::TPHandle&)+0x27) [0x55fd46f8aa47]
 15: (ThreadPool::worker(ThreadPool::WorkThread*)+0xdb9) [0x55fd4752f179]
 16: (ThreadPool::WorkThread::entry()+0x10) [0x55fd47530280]
```

for example, see /a/kchai-2017-11-17_09:57:45-rados-wip-kefu-testing-2017-11-17-1613-distro-basic-smithi/1857122

i am removing this change from my test batch.

File: src/osd/PG.cc [hunk 3]
```diff
@@ -7403,9 +7403,10 @@ boost::statechart::result PG::RecoveryState::Active::react(const MLogRec& logevt
     pg->peer_missing[logevt.from],
     logevt.from,
     context< RecoveryMachine >().get_recovery_ctx());
-  if (pg->is_peered() &&
-      got_missing)
-    pg->queue_recovery();
+  if (got_missing) {
+    post_event(DoRecovery());
```
Comment (xiexingguo) [inline src/osd/PG.cc]: Seems we still needs some sanity checking here, there are multiple substates of **Active**(e.g., **WaitRemoteRecoveryReserved**) can not handle **DoRecovery()** event properly...

**Comment by liewegas** [comment]
I forgot to add DoRecovery to the list of state events; fixed!

**Comment by gregsfortytwo** [comment]
"I'm not sure why states like Activating are paying attention tot his vevent..."

is that a question? Seems like it should be answered when fiddling with where the event is emitted.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-11-21_16:17:37-rados-wip-kefu-testing-2017-11-21-1844-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-11-22_05:41:59-rados-wip-kefu-testing-2017-11-21-1844-distro-basic-mira/

File: src/osd/PG.cc [hunk 4]
```diff
@@ -7403,9 +7403,10 @@ boost::statechart::result PG::RecoveryState::Active::react(const MLogRec& logevt
...
+    return discard_event();
```
Comment (tchaikov) [inline src/osd/PG.cc]: @liewegas we can drop this line. as `discard_event()` will always be called.

**Comment by liewegas** [comment]
fixed nit

**Comment by disaster123** [comment]
Please backport to luminous - Thanks!

File: src/osd/PG.cc (hunk 3 above)
Comment (dzafman) [reply src/osd/PG.cc]: Yes, see commit 64047e1

This issue came up in testing pull request ceph/ceph#19850 which should merge soon.

**Pull request merged**